include/ppx/knob.h

```diff
@@ -0,0 +1,389 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef ppx_knob_h
+#define ppx_knob_h
+
+#include "ppx/command_line_parser.h"
+#include "ppx/config.h"
+#include "ppx/imgui_impl.h"
+#include "ppx/log.h"
+
+#include <iostream>
+#include <unordered_set>
+#include <vector>
+
+namespace ppx {
+
+// ---------------------------------------------------------------------------------------------
+// Knob Classes
+// ---------------------------------------------------------------------------------------------
+
+// Knob is an abstract class which contains common features for all knobs and knob hierarchy
+class Knob
+{
+public:
+    Knob(const std::string& flagName)
+        : mFlagName(flagName), mDisplayName(flagName), mIndent(0) {}
+    virtual ~Knob() = default;
+
+    std::string GetDisplayName() const { return mDisplayName; }
+    std::string GetFlagName() const { return mFlagName; }
+    std::string GetFlagHelp() const { return mFlagDesc; }
+    size_t      GetIndent() const { return mIndent; }
+
+    // Populate flag members
+    void SetDisplayName(std::string displayName) { mDisplayName = displayName; }
+    void SetFlagDesc(std::string flagDesc) { mFlagDesc = flagDesc; }
+    void SetIndent(size_t indent) { mIndent = indent; }
+
+    virtual void Draw() = 0;
+
+    // Flag description in usage message
+    virtual std::string GetFlagHelpText() const = 0;
+
+    // Updates knob value from commandline flag
+    virtual void UpdateFromFlag(const CliOptions& opts) = 0;
+
+    // Default values are the values when the application starts. The knobs can be reset to default by a button in the UI
+    virtual void ResetToDefault() = 0;
+
+protected:
+    std::string mDisplayName;
+    std::string mFlagName;
+    std::string mFlagDesc;
+    size_t      mIndent;
+};
+
+// KnobCheckbox will be displayed as a checkbox in the UI
+// Currently supports variants of type: bool
+template <typename T, std::enable_if_t<std::is_same_v<T, bool>, bool> = true>
```

Comment:
The point of template is to write the same code for multiple different types, i.e. `CreateKnob<T>` and `Slider<T>`. Please don't use template when they are serve no purpose, it makes compile error harder to read and could make compile time longer.

If you feel like you need consistency
```
using KnobSliderInt = KnobSlider<int>;
using KnobSliderFloat = KnobSlider<float>;

km.CreateKnob<KnobCheckbox>(...);
km.CreateKnob<KnobSliderInt>(...);
km.CreateKnob<KnobSliderFloat>(...);
```

The usecase for template that only accept one type is usually passing to other template that needs templated types 
```
template <template<typename> DisplayT, typename ValueT, typename... ArgsT>
DisplayT<ValueT>* CreateKnob(const std::string& flagName, ArgsT... args);

km.CreateKnob<KnobCheckbox, bool>
```
and do not use complex template construct if you can help it, e.g. instead of `enable_if_t`, put a static assert in the class `static_assert(std::is_same_v<T, bool>, "Checkbox only support bool");`. Although not having them in the first place is probably a better idea.

Reply:
Yes, for the Checkbox, it seems like the template can just be dropped. As the small difference in usage between `KnobSlider<int>` vs `KnobCheckbox` doesn't justify the added complexity, as we know checkbox unless other knobs, will never take another type.